solana-programs/claimable-tokens/program/tests/tests.rs

```diff
@@ -623,18 +686,291 @@ async fn transfer_with_amount_instruction_secp_index_exploit() {
+    let tx_result = program_context
+        .banks_client
+        .process_transaction(transaction)
+        .await;
+    assert_custom_error(
+        tx_result,
+        1,
+        ClaimableProgramError::SignatureVerificationFailed,
+    )
+}
+
+// Test to verify nonce incrementing across multiple transfers
+#[tokio::test]
+async fn transfer_nonce_increment() {
+    let mut program_context = program_test().start_with_context().await;
+    let rent = program_context.banks_client.get_rent().await.unwrap();
+    let (
+        _rng,
+        _key,
+        priv_key,
+        _secp_pubkey,
+        mint_account,
+        mint_authority,
+        user_token_account,
+        eth_address,
+    ) = init_test_variables();
+
+    let mint_pubkey = mint_account.pubkey();
+    let (base_acc, user_bank_account, tokens_amount) = prepare_transfer(
+        &mut program_context,
+        mint_account,
+        rent,
+        mint_authority,
+        eth_address,
+        &user_token_account,
+    )
+    .await;
+    // Transfer a single token for reuse
+    let transfer_amount = 1;
+    let nonce_acct_seed = [NONCE_ACCOUNT_PREFIX.as_ref(), eth_address.as_ref()].concat();
+    let (_, nonce_account, _) = find_nonce_address(&id(), &mint_pubkey, &nonce_acct_seed);
+
+    let num_transfers = 5;
+    let mut i = 0;
+    let mut outgoing_account_balance = tokens_amount;
+
+    while i < num_transfers {
+        let current_user_nonce = get_user_account_nonce(&mut program_context, &nonce_account).await;
+        let transfer_instr_data = TransferInstructionData {
+            target_pubkey: user_token_account.pubkey(),
+            amount: transfer_amount,
+            nonce: current_user_nonce + 1,
+        };
+
+        let encoded = transfer_instr_data.try_to_vec().unwrap();
+        let secp256_program_instruction = new_secp256k1_instruction(&priv_key, &encoded);
+
+        let instructions = [
+            secp256_program_instruction.clone(),
+            instruction::transfer(
+                &id(),
+                &program_context.payer.pubkey(),
+                &user_bank_account,
+                &user_token_account.pubkey(),
+                &nonce_account,
+                &base_acc,
+                instruction::Transfer {
+                    eth_address,
+                    amount: transfer_amount,
+                },
+            )
+            .unwrap(),
+        ];
+
+        let mut transaction =
+            Transaction::new_with_payer(&instructions, Some(&program_context.payer.pubkey()));
+
+        transaction.sign(&[&program_context.payer], program_context.last_blockhash);
+        program_context
+            .banks_client
+            .process_transaction(transaction)
+            .await
+            .unwrap();
+
+        let final_user_nonce = get_user_account_nonce(&mut program_context, &nonce_account).await;
+        assert_eq!(transfer_instr_data.nonce, final_user_nonce);
```

Comment:
worth also assert_eq!() on the nonce value being 5?

Reply:
yep, totally agreed